lib/req_ch.ex

```diff
@@ -3,103 +3,153 @@ defmodule ReqCH do
+      "0\\n1\\n2\\n"
+
+  """
+  @spec new(Keyword.t()) :: Req.Request.t()
+  def new(opts \\ []) do
+    attach(Req.new(base_url: "http://localhost:8123"), opts)
+  end
+
+  defp attach(%Req.Request{} = req, opts) do
+    req
+    |> Req.Request.prepend_request_steps(clickhouse_run: &run/1)
+    |> Req.Request.register_options(@options)
+    |> Req.Request.merge_options(opts)
+  end
+
+  defguardp is_query_params(value) when is_list(value) or is_map(value)
+
+  @doc """
+  Performs a query against ClickHouse API.
+
+  See docs from `new/1` for details about the options.
+
+  ## Examples
+
   Queries can be performed using both `Req.get/2` or `Req.post/2`, but GET
   is "read-only" and commands like `CREATE` or `INSERT` cannot be used with it.
+  For that reason, by default we perform a `POST` request.
+  To change that, use `query/2` with a pre-configured `req`.
 
   A plain query:
 
-      iex> req = Req.new() |> ReqCH.attach()
-      iex> Req.get!(req, clickhouse: "SELECT number FROM system.numbers LIMIT 3").body
+      iex> {:ok, response} = ReqCH.query("SELECT number FROM system.numbers LIMIT 3")
+      iex> response.body
       "0\\n1\\n2\\n"
 
   Changing the format to `:explorer` will return a dataframe:
 
-      iex> req = Req.new() |> ReqCH.attach()
-      iex> Req.get!(req, clickhouse: "SELECT number FROM system.numbers LIMIT 3", format: :explorer).body
+      iex> {:ok, response} = ReqCH.query("SELECT number FROM system.numbers LIMIT 3", [], format: :explorer)
+      iex> response.body
       #Explorer.DataFrame<
         Polars[3 x 1]
         number u64 [0, 1, 2]
       >
 
   Using parameters is also possible:
 
-      iex> req = Req.new() |> ReqCH.attach(format: :explorer, database: "system")
-      iex> Req.get!(req, clickhouse: {"SELECT number FROM numbers WHERE number > {num:UInt8} LIMIT 3", [num: 5]}).body
+      iex> opts = [format: :explorer, database: "system"]
+      iex> {:ok, response} = ReqCH.query("SELECT number FROM numbers WHERE number > {num:UInt8} LIMIT 3", [num: 5], opts)
       #Explorer.DataFrame<
         Polars[3 x 1]
         number u64 [6, 7, 8]
       >
 
-  In case a different endpoint is needed, we can use the `:base_url` option
-  from `Req` to set that value. We can use use `:auth` to pass down the username
-  and password:
-
-      iex> req = Req.new(base_url: "https://example.org:8123", auth: {:basic, "user:pass"})
-      iex> req = ReqCH.attach(req)
-      iex> Req.post!(req, clickhouse: "SELECT number FROM system.numbers LIMIT 3").body
-      "0\\n1\\n2\\n"
+  """
+  @spec query(sql_query :: binary(), params :: Map.t() | Keyword.t(), opts :: Keyword.t()) ::
+          {:ok, Req.Response.t()} | {:error, binary()}
+  def query(sql_query, params \\ [], opts \\ [])
+
+  def query(sql_query, params, opts)
+      when is_binary(sql_query) and is_query_params(params) and is_list(opts) do
+    opts
+    |> new()
+    |> Req.request(clickhouse: {sql_query, params})
```

Comment:
Instead of handling it via the clickhouse option, let's just do the query and handle the `:format` directly here. No need to hook into Req's lifecycle. :)